### PR TITLE
Add certificate_model to Configuration

### PR DESCRIPTION
**Pull request by telsaiori**
- Add certificate_model for user to customize model
- Replace LetsEncrypt::Certificate to LetsEncrypt.config.certificate_model.constantize
- Update README
https://github.com/elct9620/rails-letsencrypt/issues/14?fbclid=IwAR02rN3x6sw1KdrUzYkfYEYqmIfPQtsNys7x7BmVzb1SpwWPyKxR4ljINOg

### CODE REVIEW

File: app/controllers/lets_encrypt/verifications_controller.rb [hunk 1]
```diff
@@ -17,7 +17,7 @@ def render_verification_string
     end
 
     def certificate
-      LetsEncrypt::Certificate.find_by(verification_path: filename)
+      LetsEncrypt.config.certificate_model.constantize.find_by(verification_path: filename)
```
Comment (elct9620) [inline app/controllers/lets_encrypt/verifications_controller.rb]: Prevent directly use `constantize` in each action, please implement a class method `LetsEncrypt.certificate_model` to get the model already constantize.

File: spec/controllers/lets_encrypt/verifications_controller_spec.rb [hunk 2]
```diff
@@ -12,7 +17,7 @@
 
   context 'has certificate' do
```
Comment (elct9620) [inline spec/controllers/lets_encrypt/verifications_controller_spec.rb]: The test needs to check the default certificate model and the user customizes model work correctly.

Comment (telsaiori) [reply spec/controllers/lets_encrypt/verifications_controller_spec.rb]: Added.

File: spec/controllers/lets_encrypt/verifications_controller_spec.rb [hunk 3]
```diff
@@ -4,25 +4,42 @@
...
+    context 'with default model' do
+      it 'returns verification string when found verification path' do
+        open(:get, :show, verification_path: 'valid_path')
+        expect(response.status).to eq(200)
+        expect(response.body).to eq(certificate.verification_string)
+      end
+    end
+
+    context 'with customize model' do
+      LetsEncrypt.config.certificate_model = 'OtherModel'
```
Comment (elct9620) [inline spec/controllers/lets_encrypt/verifications_controller_spec.rb]:
```ruby
before { LetsEncrypt.config.certificate_model = 'OtherModel' }
after { LetsEncrypt.config.certificate_model = 'LetsEncrypt::Certificat' }
```

If you didn't reset the certificate to default model, there have a side effect to breaks other tests.

File: spec/controllers/lets_encrypt/verifications_controller_spec.rb [hunk 4]
```diff
@@ -4,25 +4,42 @@
...
+      it 'set the certificate_model to customize model' do
+        expect(LetsEncrypt.certificate_model).to eq('OtherModel'.constantize)
```
Comment (elct9620) [inline spec/controllers/lets_encrypt/verifications_controller_spec.rb]: This test case is not related to the controller.

File: spec/controllers/lets_encrypt/verifications_controller_spec.rb [hunk 5]
```diff
@@ -4,25 +4,39 @@
...
+        expect(response.status).to eq(200)
+        expect(response.body).to eq(certificate.verification_string)
+      end
+    end
 
-    it 'returns verification string when found verification path' do
-      open(:get, :show, verification_path: 'valid_path')
-      expect(response.status).to eq(200)
-      expect(response.body).to eq(certificate.verification_string)
+    context 'with customize model' do
```
Comment (elct9620) [inline spec/controllers/lets_encrypt/verifications_controller_spec.rb]: This test didn't verify the certificate model is changed correctly.

File: spec/controllers/lets_encrypt/verifications_controller_spec.rb (hunk 4 above)
Comment (telsaiori) [reply spec/controllers/lets_encrypt/verifications_controller_spec.rb]: Moved to the  letsencrypt_spec.

File: spec/controllers/lets_encrypt/verifications_controller_spec.rb (hunk 5 above)
Comment (elct9620) [reply spec/controllers/lets_encrypt/verifications_controller_spec.rb]: You have to consider the `let!` will execute before your `context` execute `before` hook, and you still test on `LetsEncrypt::Certificate` not `OtherModel`

Comment (telsaiori) [reply spec/controllers/lets_encrypt/verifications_controller_spec.rb]: Moved let!(:certificate) to each context

Comment (elct9620) [reply spec/controllers/lets_encrypt/verifications_controller_spec.rb]: It still not **verify the certificate model is changed**, you have to change some behavior the verify it and make sure the customize behavior correctly processes.

File: lib/letsencrypt.rb [hunk 6]
```diff
@@ -86,5 +86,9 @@ def config(&block)
     def table_name_prefix
       'letsencrypt_'
     end
+
+    def certificate_model
+      config.certificate_model.constantize
```
Comment (elct9620) [inline lib/letsencrypt.rb]:
```ruby
@certificate_model ||= config.certificate_model.constantize
```

We should prevent search for the customize model every time.

File: spec/letsencrypt_spec.rb [hunk 7]
```diff
@@ -41,10 +41,13 @@
...
     class OtherModel < LetsEncrypt::Certificate
     end
-    before { LetsEncrypt.config.certificate_model = 'OtherModel' }
+    before do
+      LetsEncrypt.config.certificate_model = 'OtherModel'
+    end
     after { LetsEncrypt.config.certificate_model = 'LetsEncrypt::Certificate' }
     it 'set the certificate_model to customize model' do
-        expect(LetsEncrypt.certificate_model).to eq('OtherModel'.constantize)
+      allow(LetsEncrypt).to receive(:certificate_model).and_return('OtherModel'.constantize )
```
Comment (elct9620) [inline spec/letsencrypt_spec.rb]: `allow` didn't do any `assert` this test didn't verify anything.

File: spec/letsencrypt_spec.rb [hunk 8]
```diff
@@ -37,4 +37,17 @@
       LetsEncrypt.register('example@example.com')
     end
   end
+
+  describe 'certificate_model' do
+    class OtherModel < LetsEncrypt::Certificate
+    end
+    before do
+      LetsEncrypt.config.certificate_model = 'OtherModel'
+      LetsEncrypt.stub(:certificate_model) { 'OtherModel'.constantize }
```
Comment (elct9620) [inline spec/letsencrypt_spec.rb]:
```ruby
LetsEncrypt.stub(:certificate_model) { LetsEncrypt.config.certificate_model.constantize }
 ```

The stub method returns `'OtherModel'.constantize` but your assert is `'OtherModel'.constantize` too, they're no different and didn't check anything.
